pkg/kubectl/cmd/cmd.go

```diff
@@ -598,6 +601,57 @@ func NewKubectlCommand(in io.Reader, out, err io.Writer) *cobra.Command {
 		}
 	}
 
+	cmdutil.CheckErr(cmds.RegisterFlagCompletionFunc("namespace", func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		clientSet, err := f.KubernetesClientSet()
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveDefault
+		}
+		ctx, cancel := context.WithTimeout(context.TODO(), time.Second*3)
+		defer cancel()
+		namespaces, err := clientSet.CoreV1().Namespaces().List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return nil, cobra.ShellCompDirectiveDefault
+		}
+		var ret []string
+		for _, ns := range namespaces.Items {
+			if strings.HasPrefix(ns.Name, toComplete) {
+				ret = append(ret, ns.Name)
+			}
+		}
+		return ret, cobra.ShellCompDirectiveNoFileComp
+	}))
+
+	getContextFromConfig := getFromConfig(f, func(config clientcmdapi.Config, toComplete string) []string {
+		var ret []string
+		for name := range config.Contexts {
+			if strings.HasPrefix(name, toComplete) {
+				ret = append(ret, name)
+			}
+		}
+		return ret
+	})
+	getClusterFromConfig := getFromConfig(f, func(config clientcmdapi.Config, toComplete string) []string {
+		var ret []string
+		for name := range config.Clusters {
+			if strings.HasPrefix(name, toComplete) {
+				ret = append(ret, name)
+			}
+		}
+		return ret
+	})
+	getUserFromConfig := getFromConfig(f, func(config clientcmdapi.Config, toComplete string) []string {
+		var ret []string
+		for name := range config.AuthInfos {
+			if strings.HasPrefix(name, toComplete) {
+				ret = append(ret, name)
+			}
+		}
+		return ret
+	})
+	cmdutil.CheckErr(cmds.RegisterFlagCompletionFunc("context", getContextFromConfig))
+	cmdutil.CheckErr(cmds.RegisterFlagCompletionFunc("cluster", getClusterFromConfig))
+	cmdutil.CheckErr(cmds.RegisterFlagCompletionFunc("user", getUserFromConfig))
```

Comment:
It seems like there is going to be a lot of code here as you add more completions.  What about putting all the completion code into its own separate package and then make a one-line callout from here to register completions?

Reply:
updated